File: example/src/com/appbrain/example/fun/SpringObject.java
package com.appbrain.example.fun;

public class SpringObject {
    public static final float springStrength = 0.1f;
    public static final float friction = 0.8f;
    public Vector pos, vel, goal;

    public SpringObject(Vector start) {
        this.pos = new Vector(start);
        this.vel = new Vector(0, 0);
        this.goal = new Vector(start);
    }

    private Vector d = new Vector();

    public void update() {
        Vector.sub(goal, pos, d);
        d.mult(springStrength);
        vel.add(d);
        vel.mult(friction);
        pos.add(vel);
    }
}

File: example-compose/src/com/appbrain/example/logo/SpringObject.kt
package com.appbrain.example.logo

open class SpringObject(start: Vector) {
    val pos: Vector = Vector(start)
    private val vel: Vector = Vector(0f, 0f)
    val goal: Vector = Vector(start)

    private val d = Vector()

    open fun update() {
        Vector.sub(goal, pos, d)
        d.mult(SPRING_STRENGTH)
        vel.add(d)
        vel.mult(FRICTION)
        pos.add(vel)
    }

    companion object {
        const val SPRING_STRENGTH = 0.1f
        const val FRICTION = 0.8f
    }
}
